### ISSUE TITLE
Add documentation for Clustermesh setup through Helm Chart

### ISSUE THREAD

**Issue by kaworu** [cilium/cilium#19057]
Follow-up issue of https://github.com/cilium/cilium/pull/17851 which introduced support to connect multiple Clustermesh clusters using the Helm Chart.

**Comment by samueltorres** [comment]
hey @aanm I'm currently working on it ! 🚀

**Comment by datsabk** [comment]
FWIW, I tried to set this up and faced bunch of difficulties. List out the roadblocks here: 
1. We need to manually enter details of every cluster we want to connect 
2. The API Server CA needs to be manually entered without which the configuration wouldn't go through 
3. The DNS records are not created by Helm - So we need an add-on like `external-dns` that can create records based on the provided annotation.

**Comment by xinity** [comment]
@datsabk any chances you could share details on how you achieved the clustermesh setup through helm chart ?

**Comment by datsabk** [comment]
@xinity Wasn't able to. There is too much of certificate configuration complexity and some values are just not being accepted. The approach i took was - use cilium cli to install and connect two clusters, gather the parameters from the live install and try to use that in helm chart.

**Comment by dgiebert** [comment]
I am currently trying to achieve the cluster mesh with helm, and this needs to be a two-step process, in my opinion:

For this example the final setup will be cluster01 <-> cluster02

1. Enable the API Server on both clusters that you want to join the mesh (Untested and constructed from HelmChartConfig below)
```sh
helm upgrade cilium cilium/cilium --version 1.12.3 \
   --namespace kube-system \
   --reuse-values \
   --set ipv4NativeRoutingCIDR=10.0.0.0/9 \
   --set kubeProxyReplacement=strict \
   --set l7Proxy=false \
   --set k8sServiceHost=x.x.x.x \
   --set k8sServicePort=6443 \
   --set cluster.name=cluster01 \
   --set cluster.id=1 \
   --set externalWorkloads.enabled=true \
   --set clustermesh.useAPIServer=true
```
2. Extract the CA, Cert, and Key for every cluster from the secret `clustermesh-apiserver-remote-cert` in kube-system
2. Create the link on both cluster with the needed values (Untested and constructed from HelmChartConfig below)
```sh
helm upgrade cilium cilium/cilium --version 1.12.3 \
   --namespace kube-system \
   --reuse-values \
   --set ipv4NativeRoutingCIDR=10.0.0.0/9 \
   --set kubeProxyReplacement=strict \
   --set l7Proxy=false \
   --set k8sServiceHost=x.x.x.x \
   --set k8sServicePort=6443 \
   --set cluster.name=cluster01 \
   --set cluster.id=1 \
   --set externalWorkloads.enabled=true \
   --set clustermesh.useAPIServer=true \
   --set clustermesh.config.enabled=true \
   --set clustermesh.config.clusters[0].name=cluster02 \
   --set clustermesh.config.clusters[0].ips[0]=x.x.x.x \
   --set clustermesh.config.clusters[0].port=32379 \
   --set clustermesh.config.clusters[0].tls.cert=.... \
   --set clustermesh.config.clusters[0].tls.key=.... \
   --set clustermesh.apiserver.tls.ca.cert=.... \
```

This actually done using rke2, here is the `HelmChartConfig` for it:
```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-cilium
  namespace: kube-system
spec:
  valuesContent: |-
    kubeProxyReplacement: strict
    k8sServiceHost: x.x.x.x
    k8sServicePort: 6443
    ipv4NativeRoutingCIDR: 10.0.0.0/9
    l7Proxy: false
    cluster:
      name: cluster01
      id: 1
    externalWorkloads: 
      enabled: true
    clustermesh:
      useAPIServer: true
      config:
        enabled: true
        clusters:
        - name: cilium04
          ips:
          - x.x.x.x
          port: 32379
          tls:
            cert: ...
            key: ...
      apiserver:
        tls:
          ca:
            cert: ....

```

This will for now only work with two clusters as the CA is not configurable per cluster. 
@samueltorres what was the design decision for this?

**Comment by datavisorzhizhu** [comment]
@dgiebert where is **clustermesh-apiserver-remote-cert** being used?

**Comment by dgiebert** [comment]
On cilium01 (and on the other cluster you will need to adapt that accordingly):
```
        clusters:
        - name: cilium02
          ips:
          - x.x.x.x
          port: 32379
          tls:
            cert: "Check clustermesh-remote-server-cert on kube-system in cluster02"
            key: "Check clustermesh-remote-server-cert on kube-system in cluster02"
        - name: cilium03
          ips:
          - x.x.x.x
          port: 32379
          tls:
            cert: "Check clustermesh-remote-server-cert on kube-system in cluster03"
            key: "Check clustermesh-remote-server-cert on kube-system in cluster03"
```


Using helm they should be set in the array:
```
   --set clustermesh.config.clusters[0].name=cluster02 \
  [...]
   --set clustermesh.config.clusters[0].tls.key=.... \
   --set clustermesh.config.clusters[1].name=cluster03 \
  [...]
   --set clustermesh.config.clusters[1].tls.key=.... \
```
Keep in mind that I did only test it using HelmChartConfig!

**Comment by datavisorzhizhu** [comment]
thanks a lot

**Comment by darkstarmv** [comment]
I tried to follow [cilium/clustermesh-apiserver/README.md#deploy-using-helm](https://github.com/cilium/cilium/blob/master/clustermesh-apiserver/README.md#deploy-using-helm) steps, but my clustermesh can't initialize:

**Comment by datavisorzhizhu** [comment]
@darkstarmv you can follow this blog: https://www.sobyte.net/post/2022-05/cilium-cluster-mesh/ using cilium command;
when you understand steps for setting up cilium cluster mesh, you can convert the steps into helm chart based on the blog and above discussion

**Comment by m-yosefpor** [comment]
still relevant

**Comment by arunvelsriram** [comment]
Am using Terraform to manage K8s resources. For clustermesh setup I took a two step approach:

1. cilium-cni installation: creates cilium release in each cluster and stores the clustermesh remote connection details as output in the Terraform state ([sensitive output](https://developer.hashicorp.com/terraform/language/values/outputs#sensitive-suppressing-values-in-cli-output))
2. clustermesh connection: gets the clustermesh remote connection details from the Terraform state for each cluster and pushes them in to the current cluster

Rough implementation:

1. cilium-cni installation

```terraform

# cilium helm release
resource "helm_release" "cilium" {
...
}

# read the clustermesh remote connection certs and api server loadbalancer ip
data "kubernetes_service_v1" "clustermesh_apiserver" {
  metadata {
    name      = "clustermesh-apiserver"
    namespace = "kube-system"
  }

  depends_on = [helm_release.cilium]
}

data "kubernetes_secret_v1" "clustermesh_remote_cert" {
  metadata {
    name      = "clustermesh-apiserver-remote-cert"
    namespace = "kube-system"
  }

  depends_on = [helm_release.cilium]
}

# output the remote connection details
locals {
  clustermesh_apiserver_ip          = data.kubernetes_service_v1.clustermesh_apiserver.status[0].load_balancer[0].ingress[0].ip
  clustermesh_apiserver_port        = data.kubernetes_service_v1.clustermesh_apiserver.spec[0].port[0].port
  clustermesh_apiserver_endpoint    = "https://${local.clustermesh_apiserver_ip}:${local.clustermesh_apiserver_port}"
  clustermesh_apiserver_remote_cert = data.kubernetes_secret_v1.clustermesh_remote_cert.data
}

output "clustermesh_remote_connection" {
  value = {
    endpoints = [local.clustermesh_apiserver_endpoint]
    cert      = local.clustermesh_apiserver_remote_cert
  }
  description = "Connection details for other peers in the mesh to connect securely to the clustermesh apiserver"
  sensitive   = true
}
```

2. clustermesh connection

```terraform
# read clustermesh remote connection details from tfstate of each cluster
data "terraform_remote_state" "clustermesh_connection" {
  for_each = var.cilium_cni_tf_states
  backend  = "local"

  config = {
    path = each.value
  }
}

locals {
  data = [
    for cluster, connection in data.terraform_remote_state.clustermesh_connection.outputs : {
      (cluster) = templatefile("${path.module}/templates/etcd_config.tftpl", {
        cluster = cluster, endpoints = lookup(connection, "endpoints")
      })
      "${cluster}-ca.crt" = lookup(lookup(connection, "cert"), "ca.crt")
      "${cluster}.crt"    = lookup(lookup(connection, "cert"), "tls.crt")
      "${cluster}.key"    = lookup(lookup(connection, "cert"), "tls.key")
    } if cluster != var.cluster_name # skip current cluster's connection details
  ]
}

resource "kubernetes_secret_v1" "clustermesh_connection" {
  metadata {
    name      = "cilium-clustermesh"
    namespace = "kube-system"
  }

  data = merge(local.data...)
}
```

**templates/etcd_config.tftpl**
> This is not yaml I guess. The content should be exactly same otherwise the cilium-agent was not able to read the data. I faced issues when I had double quotes around keys and values. Reference from `cilium-cli`: https://github.com/cilium/cilium-cli/blob/main/clustermesh/clustermesh.go#L662-L670
```tftpl
endpoints:
%{ for endpoint in endpoints ~}
- ${endpoint}
%{ endfor ~}
trusted-ca-file: /var/lib/cilium/clustermesh/${cluster}-ca.crt
cert-file: /var/lib/cilium/clustermesh/${cluster}.crt
key-file: /var/lib/cilium/clustermesh/${cluster}.key
```

**Comment by arunvelsriram** [comment]
Also, [The instructions mentioned in readme](https://github.com/cilium/cilium/blob/main/clustermesh-apiserver/README.md#connect-cilium-clusters-in-to-a-clustermesh-1) worked for me. I just modified https://github.com/cilium/cilium/blob/main/contrib/k8s/k8s-extract-clustermesh-nodeport-secret.sh script to extract clustermesh-api loadbalancer ip instead of node ip.

**Comment by asauber** [comment]
The cilium-cli now defaults to Helm mode, in which `clustermesh {enable,disable}` are implemented by [modifying helm values](https://github.com/cilium/cilium-cli/blob/main/clustermesh/clustermesh.go#L1911) in the cluster.

With Cilium 1.14+ and the standard Clustermesh documentation, you will have a ClusterMesh setup via Helm: https://docs.cilium.io/en/stable/network/clustermesh/clustermesh/#clustermesh

**Comment by dazmc** [comment]
> Also, [The instructions mentioned in readme](https://github.com/cilium/cilium/blob/main/clustermesh-apiserver/README.md#connect-cilium-clusters-in-to-a-clustermesh-1) worked for me. I just modified https://github.com/cilium/cilium/blob/main/contrib/k8s/k8s-extract-clustermesh-nodeport-secret.sh script to extract clustermesh-api loadbalancer ip instead of node ip.

@arunvelsriram what changes did you make to use loadbalancer ip?

**Comment by project-administrator** [comment]
Can anyone please post a complete helm values example for a working cluster mesh setup?

**Comment by mayur2281** [comment]
Hi I am still facing certificate issues wile trying to configure clustermesh between 2 clusters using helm and ArgoCD. Cant we have a init script to get the necessary certificate secrets in sync with both the clusters?

**Comment by mayur2281** [comment]
https://github.com/cilium/cilium/issues/32076#issuecomment-2069415020

Here is the difference in cli install and helm install.

**Comment by mayur2281** [comment]
Hi @all Should i add the secrets manually like this as explained in this blog: 

https://docs.kubermatic.com/kubermatic/v2.25/tutorials-howtos/networking/cilium-cluster-mesh/

Is there a way to automate this using helm charts itself?

**Comment by dazmc** [comment]
Note: I don't use clustermesh.apiserver.tls..ca  and clustermesh.apiserver.config.clusters[].tls in favour of setting it in tls.ca

@mayur2281 You can create the cert/key and base64 encoded it and then have it in the helm values for both clusters. I normally just deploy the first cluster export the certs and then keep it safe (keystore). Then as i build other clusters i put the base64 encoded values in tls.ca in helm values file

**Comment by mayur2281** [comment]
@dazmc When you deploy the first cluster you export the secret called cilium-ca right? and then put that tls.ca section? Other than that there is no other configuration?

**Comment by mayur2281** [comment]
@samueltorres Can we have another section for clustermesh helm install in these docs?: https://docs.cilium.io/en/stable/network/clustermesh/clustermesh/#gs-clustermesh

**Comment by mayur2281** [comment]
@dazmc Pls tell me what secrets from the first cluster are to be used in the second cluster and in which configuration blocks in the helm values! It is really confusing. Are you suggesting to use only this block here: https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/values.yaml#L2248

**Comment by mayur2281** [comment]
Thanks @dazmc Got the clustermesh working by having the same CA as the first cluster in my second cluster in the tls.ca section in the helm values.

**Comment by dazmc** [comment]
@mayur2281  yes, that's correct. If you notice [here](https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/values.yaml#L2857C1-L2859C47)  it says "These fields can (and should) be omitted in case the CA is shared across clusters" and yes it is confusing ;-)

**Comment by mayur2281** [comment]
@dazmc Yes i saw that and was clear not to use that section, but what was confusing for me was the clustermesh.apiserver.tls section. This section is purely used for communication within the cluster for the clustermesh apiserver right?

Anyways i am writing a blog regarding this and will share it here, Pls share your review on it!

**Comment by dazmc** [comment]
@mayur2281 i then have key/cert in a files (don't do this in production) and use envsubst to fill a template helm values file, including other variables.

```
#mesh_helm.tmpl
.....
ipv4NativeRoutingCIDR: ${CILIUM_CIDR}
cluster:
  name: CLUSTER1
  id: 1
tls:
  ca:
    cert: $CA_CERT
    key: $CA_KEY
```

```
export CA_CERT=$(cat /path/to/CA_FILE)
export CA_KEY=(cat /path/to/CA_KEY)
 cat mesh_helm.tmpl | envsubst > mesh_helm.yaml
```

**Comment by mayur2281** [comment]
https://github.com/cilium/cilium/issues/19057#issuecomment-2072103666

I cant do this because i'm using ArgoCD to deploy the clustermesh and have to insert the cert and key manually in the helm values and pass it to the second cluster in argocd.

**Comment by dazmc** [comment]
@mayur2281 there is nothing stopping you using the cert/key on both cluster1 helm values as well as cluster2. I generated that first cluster to get the cert/key format then i blew it away. Now i have the cert/key so i just put it in the helm values when i deploy cluster 1 or cluster 2 etc... 

You can manually create the cert/key and use it on any cluster (1, 2, 3 etc...) . So all my clusters have the same key. Sometimes i generate a new cert/key so then i can have cluster1,2,3 in one mesh and cluster4,5,6 in a different mesh. Just keep the cert/keys you need/want.  

Note: I don't use ArgoCD I prefer flux ;-)

**Comment by mayur2281** [comment]
@dazmc So then what would be the best practice for Production?

**Comment by dazmc** [comment]
@mayur2281  Having the ca key/certs laying around in a file on your laptop isn't a good idea imho. So i would at least encrypt the files (sops) but better still store the cert/key in something like hashicorp vault or external-secrets (AWS /Azure/GCP keyvault).

**Comment by mayur2281** [comment]
Yes will do that

**Comment by mayur2281** [comment]
@dazmc Could you review this? https://notes.mayurbn.site/Blogs/Cilium-Clustermesh-Deployment-on-Helm-or-ArgoCD-on-AKS

**Comment by samueltorres** [comment]
Hey folks,

Just created an example repo on how to configure clustermesh through the helm chart:
https://github.com/samueltorres/cilium-clustermesh-helm/

I aim in the future as soon as I have some free time to document the clustermesh through Helm docs :)

**Comment by dazmc** [comment]
@mayur2281 I would swap lines 12 and 13 around in first yaml and lines 17 and 18 around in the second yaml. 

You also don't mention how/where you get the load balancer addresses in the yamls

**Comment by mayur2281** [comment]
@dazmc I have made the changes. Pls take a look at it again and share the feedback.
https://notes.mayurbn.site/Blogs/Cilium-Clustermesh-Deployment-on-Helm-or-ArgoCD-on-AKS

**Comment by dazmc** [comment]
@mayur2281 corporate firewall blocks your website so will have to check later.

@samueltorres nice,  I tend to set clustermesh.apiserver.tls.auto.enabled to true
```
  apiserver:
    tls:
      auto:
        enabled: true
```
Then all the certs are autogenerated for admin, server, client etc... [here](https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/values.yaml#L3082-L3086) for mTLS. Any reason you set it to false and provide the certs manually? just interested in another viewpoint

**Comment by mayur2281** [comment]
Its set to false because he creates all his certs through the shell script.

**Comment by dazmc** [comment]
@mayur2281 yes i understand, that but I'm trying to understand why you wouldn't just leave it set to auto, then those admin, server, client are autogenerated. it seems like more work as you have to prepare create/store/manage them. I'm sure there is a reason that @samueltorres  has done it that way.

**Comment by mayur2281** [comment]
Oh yes Surely he'd have his reasons

**Comment by samueltorres** [comment]
In my case I'd be leveraging Vault as a [PKI store](https://developer.hashicorp.com/vault/docs/secrets/pki), so I tried to emulate the same behaviour here. In my company we install Cilium through Terraform so this is all easy to do from within our Terraform configuration.

**Comment by mayur2281** [comment]
@dazmc I would like to store the certs in Azure Key Vault and use a Secret store CSI driver to create a secret from there. But there is no option in the values to provide a secret name all I have is tls.ca.cert and tls.ca.key.

**Comment by mayur2281** [comment]
@samueltorres Any suggestions on the above comment?
